content/blog/grpc-web-interceptor.md

```diff
@@ -1,46 +1,30 @@
 ---
-title: Using gRPC-Web Interceptors
-date: 2020-06-01
+title: Interceptors in gRPC-Web
+date: 2020-06-11
 authors:
 - name: Zhenli Jiang
   position: Google
 ---
 
-Interceptor is a very common pattern for RPC client APIs in many languages.
-gRPC-Web has also released this feature recently (May 2020 v1.1.0). The current
-design is based on other gRPCs’ (Java, Node, Go, etc) client interceptors, with
-some gRPC-Web specific features for better developer experience. Meanwhile,
-we’ve made a lot of effort to integrate gRPC-Web and the client interceptor
-design into Google’s internal Web frameworks. We believe that our current design
-allows gRPC-Web to be easily integrated into modern Web frameworks and the API
-is also very easy for gRPC-Web users to adopt.
+We're pleased to announce support for _interceptors_ in [gRPC-web][] as of
+release [1.1.0][]. While the current design is based on gRPC client interceptors
+available from other [gRPC languages][], it also includes gRPC-web specific
+features that should make interceptors easy to adopt and use alongside modern
+web frameworks.
 
 ## Introduction
 
-In [gRPC-Web](https://github.com/grpc/grpc-web) there are two kinds of [client
-interceptors](https://github.com/grpc/grpc-web/blob/master/javascript/net/grpc/web/interceptor.js).
-The type of interceptors is determined by the response type of the RPC call.
+Similar to other gRPC languages, gRPC-web supports _unary_ and server
+_streaming_ interceptors. For each kind of interceptor, we've [defined an
+interface][interceptor.js] containing a single `intercept()` method:
 
-- **UnaryInterceptor**: for Promise responses
-- **StreamInterceptor**: for ClientReadableStream responses
+- `UnaryInterceptor`
+- `StreamInterceptor`
 
-gRPC-Web has two clients generated by
-[Protoc](https://github.com/grpc/grpc-web#client-configuration-options) for one
-service: let’s say *FooServiceClient* and *FooServicePromiseClient*. The
-response type of *FooServiceClient* is
-[ClientReadableStream](https://github.com/grpc/grpc-web/blob/master/javascript/net/grpc/web/clientreadablestream.js)
-for BOTH unary calls and server streaming calls, so **StreamInterceptor** is
-expected to be used for intercepting *FooServiceClient* calls. The response type
-of *PromiseClient* is
-[Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise),
-so we should use **UnaryInterceptor** for *PromiseClients*.
```

Comment:
FYI: these are the details of the use of `protoc` that I've dropped from this revision.

Reply:
I do want to add some details on the use case of StreamInterceptor and UnaryInterceptor in this blog, since it's a very confusing concept. : (
Users will (I guess) usually think StreamInterceptor is used for streaming calls and UnaryInterceptor is for unary calls. But actually StreamInterceptor can intercept both unary and streaming calls.  Which interceptor to use is only decided by the client type they are using.
I'll also add this to the code comments as well

Reply:
Ok. How about I add a note to this effect -- i.e., that a StreamInterceptor can be used to intercept both unary and streaming calls.

(Do you have an example of a such a StreamInterceptor?)

Reply:
Actually, investigating this more, while I see that the grpc-web protoc plugin generates two clients for each service, this is nonstandard AFAIK. @stanley-cheung Is this documented anywhere?

To say that a StreamInterceptor can be used to intercept both unary and streaming calls seems misleading for someone who might be familiar with gRPC but not the particularities of grpc-web.

Ok, I'll need to give this some more thought. It might be clearer to say that a StreamInterceptor can be used for any RPC with a ClientReadableStream return type (which would be either a server-streaming RPCs and unary RPCs from the generated client that only uses ClientReadableStream for return types (is there a name for this kind of generated client?).

Reply:
Yes currently the grpc-web protoc plugin generate both the normal callback-based client (which returns ClientReadableStream for both unary and streaming calls), as well as the Promise-based client which wasn't very well documented. I can see how it's difficult to convey this in this blog post. To further complicate things, the Promise-based client is only promise-based for unary calls (there is no promise-based streaming API), hence the difficult naming of `UnaryInterceptor`.

I think it will be wise to emphasize that `StreamInteceptor` is for calls that return `ClientReadableStream`.

But on the other hand, perhaps should we consider to _only_ focus this blog post on `StreamInterceptor`? This way our message can be more focused and less confusing. Perhaps we can leave Promise-based client as a totally separate topic for a later time?

Reply:
> I think it will be wise to emphasize that StreamInteceptor is for calls that return ClientReadableStream.

Agreed.

> ... perhaps should we consider to only focus this blog post on StreamInterceptor

That might be more effort at this point, especially if we agree to simply emphasize that: "a `StreamInteceptor` can be applied to any call with a `ClientReadableStream` return type (whether a unary or a server-streaming RPC)"

How does that sound?

Reply:
Sounds good to me. Thanks!

Reply:
Added a note.